README.org

```diff
@@ -265,10 +268,68 @@ releases. Use the ~fill~ command instead.
 report that it filled in more than one block. This is expected, and will occur
 when orphans/forks are present at that height.
 
+*** migrate
+
+~migrate~ allows you to migrate the database schema to the latest version and exit.
+This can be useful for separating the migration step from running the ETL and/or HTTP service.
+
+#+begin_example
+  > chainweb-data migrate --dbuser=joe --dbname=chainweb-data
+#+end_example
+
+
 *** check-schema
 
 ~check-schema~ is used to perform a check of the ORM definitions against the DB schema.
 
 #+begin_example
   > chainweb-data check-schema --service-host=foo.chainweb.com --dbuser=joe --dbname=chainweb-data
 #+end_example
+
+** Specializing the Database Schema
+
+A common use case for ~chainweb-data~ is to primarily run it as an ETL process to
+populate a Postgres database with the blockchain data.  In this case, ~chainweb-data~
+operators often want to run their schema migrations and specialize the schema according
+to their own needs. Obviously, ~chainweb-data~ can not guarantee unimpeded operation
+under arbitrary schema changes, so any node operator that wishes to specialize the
+database takes on the responsibility of ensuring that their extra schema changes
+do not interfere with the operation of ~chainweb-data~ at the time of defining those
+changes and also in the face of schema changes that will be introduced by future
+~chainweb-data~ versions.
+
+That said ~chainweb-data~ does support a structure in order to help with this process.
+Any version of ~chainweb-data~ comes with a set of schema migrations included in the
+binary that are applied by default to the database at migration time. These migrations
+are defined in the ~haskell-src/db-schema/migrations~ directory. It is possible to override
+these migrations by calling ~chainweb-data~ with the optional ~--migrations-folder~ argument.
+However, in order to facilitate the typical use case, which is to add migrations to the
+default set, an ~--extra-migrations-folder~ argument is also provided.
+
+The default migrations that come with ~chainweb-data~ generally have the following file
+name format: ~X.Y.Z.N_NAME.sql~, where ~X.Y.Z~ is the version of ~chainweb-data~ that the
+migration was introduced right after and ~N~ is the migration number. These migrations are executed
+in "dictionary order" if one considers the numeric values of X,Y,Z and N to be the letters.
+The migration procedure will scan the names and the content hashes of the existing migrations
+and expect them to be coherent with the migrations provided via the ~--extra-migrations-folder~
+and ~--migrations-folder~ arguments (or the default migrations included in the binary).
```

Comment:
```suggestion
The migration procedure will verify the migrations provided through the argument ~--extra-migrations-folder~, and check if they're coherent with the existing migrations (provided in the binary or via the ~--migrations-folder~ argument).
```

Reply:
Your version makes it sound like there's a comparison between the migrations in `--extra-migrations-folder` and `--migrations-folder`, while in reality, We compare (`--extra-migrations-folder` + `--migrations-folder`) against the existing migrations we find in the database that we connect to. One thing I wanted to convey in this paragraph is that CW-D is very picky about the consistency between desired migrations vs. existing migrations in the database. The existing migrations need to have exactly the same contents and they need to have been run in exactly the same order without any gaps. If CW-D detects any violation of any of this, it will bail out.

Reply:
Maybe we can add something along the lines of: and checks with already executed migrations in the database

Reply:
That's a good point, reading back my version, it's not clear what "existing migrations" refers to. Then how about:

```suggestion
The migration procedure will fetch the already executed migrations from the database and check them against the migrations provided through the ~--extra-migrations-folder~ and the ~--migrations-folder~ arguments.
```